### ISSUE TITLE
Synchronous vs Asynchronous

### ISSUE THREAD

**Issue by sm2017** [auth0/node-jsonwebtoken#566]
# Synchronous 
```js
var jwt = require('jsonwebtoken');
var token = jwt.sign({ foo: 'bar' }, 'shhhhh');
```

# Asynchronous
```js
var token = jwt.sign({ foo: 'bar' }, 'shhhhh', function(err, token) {
  console.log(token);
);
```

I want to know , is it bad to use Synchronous  way , in express js server? is there any benefit with Asynchronous mode? As I think there is no I/O and just CPU usage, So Asynchronous and Synchronous must b same, just syntax is different

**Comment by sm2017** [comment]
@ziluvatar , @MitMaro  can you help me?

**Comment by sm2017** [comment]
Hello???????????

**Comment by sm2017** [comment]
:-((

**Comment by ziluvatar** [comment]
> in point of performance there is no different I use sync or async mode, and it is safe to use sync mode in server too

Correct. Although if you use the sync way be sure you wrap it with try-catch.

**Comment by lonix1** [comment]
If there is no perf difference, why is there an async alternative?

**Comment by ziluvatar** [comment]
TL;DR; To allow you load keys from filesystem or any other source of [ReadStream](https://nodejs.org/docs/latest-v6.x/api/fs.html#fs_class_fs_readstream) without blocking event loop. Not to actually perform the signing.

I've been digging into git history and jws implementation to understand why the library provided the async functionality for signing:
* Back in 2013, `jws` implemented streaming for sign/verify: https://github.com/brianloveswords/node-jws/commit/ad80a8ac9fdea2de59b607ea8f556451053af145
* In 2015 that was used by this library https://github.com/auth0/node-jsonwebtoken/pull/127 , however, in my opinion, they misunderstood "streaming support" with "async signing".
* `jws` library provides streaming for payload and secret, that way you could load those from the file system or any source that provided a NodeJS stream interface, like we can see in `jws`: https://github.com/brianloveswords/node-jws/blob/ad80a8ac9fdea2de59b607ea8f556451053af145/test/jws.test.js#L132
* On this library we supported both streams until v7.4.3 -> v8.0.0 where we lost compatibility with stream payload ( not known breaking change 🎉 😞, I just updated the [migration notes v7 -> v8](https://github.com/auth0/node-jsonwebtoken/wiki/Migration-Notes:-v7-to-v8) ).

Honestly, not sure if that streaming is a desired feature by consumers, we'd need to update the README to make it clear.

**Comment by lonix1** [comment]
@ziluvatar That sounds like some serious coding archaeology... well done! :wink:
